### PR TITLE
Handling and tests for expression trees

### LINKED ISSUE #3: `is null` and `is not null` can't be used in expression contexts
Looking at https://github.com/AArnott/CSharpIsNull/blob/main/src/CSharpIsNullAnalyzer/CSIsNull001.cs and https://github.com/AArnott/CSharpIsNull/blob/main/src/CSharpIsNullAnalyzer/CSIsNull002.cs, I didn't see anything checking for this.

```cs
using System;
using System.Linq.Expressions;

// ❌ CS8122 An expression tree may not contain an 'is' pattern-matching operator.
//                                        ↓↓↓↓↓↓↓↓↓
_ = (Expression<Func<string, bool>>)(s => s is null);

// ❌ CS8122 An expression tree may not contain an 'is' pattern-matching operator.
//                                        ↓↓↓↓↓↓↓↓↓↓↓↓↓
_ = (Expression<Func<string, bool>>)(s => s is not null);

_ = (Expression<Func<string, bool>>)(s => s is object); // Works
```

[(SharpLab)](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCeAfAAQCYBGAWACgCAGAAgJIDoAZWAR0YFEAPAB3VQBnQVAD2MQQG5KlAPSzagGXJaAYQDKADhJEitAIIxaqPgOFjDCAalqQctGKIS0AxuITRDEQwHIog77S8EAgIGDAAtJAIzgAWsADmtKK8GMGi6IxyCrQ5uXn5BYVFuYDJhGXl5ZQA+rQAvLQAFDz8QiLiADwEAKydJNRwtMCiogA2AHxjAJQNgnVjtLN+9gCuIyOT0hRZSqqa2roGRiat5rSWqNa29o4ubh60XrS+/oHBoegRUbEJSSnoaRltsVgSDihVwRVqnVGs1TG0YJ0egx+oNhuMpjM5gtaEsHE4YKt1psofUmsczB1ur0UUNRhNprNavNFrNRMAAFaoZwIDa0eS0ADq6QA1oJKEA)

### PR DESCRIPTION
**Pull request by AArnott**
These tests repro the cases called out in #3, plus a couple more.

Fixes #3

### CODE REVIEW

File: src/CSharpIsNullAnalyzer/CSIsNull001.cs [hunk 1]
```diff
@@ -44,6 +44,12 @@ public override void Initialize(AnalysisContext context)
             context.EnableConcurrentExecution();
             context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.ReportDiagnostics);
 
+            context.RegisterOperationAction(
+                ctxt =>
+                {
+                    // ctxt.Operation.
```
Comment (AArnott) [inline src/CSharpIsNullAnalyzer/CSIsNull001.cs:50]: @sharwell do you have any suggestions for detecting whether an expression is part of an expression tree? I imagine we need more than just the syntax tree, so I've looked at using IOperation, but it's not clear to me how to use that API either.

File: test/CSharpIsNullAnalyzer.Tests/CSIsNull002Tests.cs [hunk 2]
```diff
@@ -177,4 +177,66 @@ void Method(object o)
...
+class Test
+{
+    void Method()
+    {
+        Expression<Func<string, bool>> e = s => s is object;
+    }
+}";
+
+        await VerifyCS.VerifyCodeFixAsync(source, fixedSource, CSIsNull002Fixer.IsObjectEquivalenceKey);
+        // TODO: assert that a IsNotNullEquivalenceKey is NOT offered, as that would produce a compilation error.
```
Comment (AArnott) [inline test/CSharpIsNullAnalyzer.Tests/CSIsNull002Tests.cs:209]: @sharwell How do I use your analyzer framework to assert that a particular code fix (by equivalence key) is _not_ offered?

Comment (sharwell) [reply test/CSharpIsNullAnalyzer.Tests/CSIsNull002Tests.cs:209]: Set `CodeActionEquivalenceKey` to the item which should not be offered, and set `TestCode` and `FixedCode` to the same content. When `TestCode` and `FixedCode` are the same, the default number of expected iterations is 0, which means no matching code action is presented. Setting `CodeActionEquivalenceKey` filters the candidate actions:

https://github.com/dotnet/roslyn-sdk/blob/e5420e26d42234bb189aa43f093fc7d79e2a2bee/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/CodeActionTest%601.cs#L138

File: src/CSharpIsNullAnalyzer/CSIsNull001.cs (hunk 1 above)
Comment (sharwell) [reply src/CSharpIsNullAnalyzer/CSIsNull001.cs:50]: https://github.com/dotnet/roslyn-analyzers/blob/762b08948cdcc1d94352fba681296be7bf474dd7/src/Utilities/Compiler/Extensions/IOperationExtensions.cs#L519-L522

File: test/CSharpIsNullAnalyzer.Tests/CSIsNull002Tests.cs (hunk 2 above)
Comment (AArnott) [reply test/CSharpIsNullAnalyzer.Tests/CSIsNull002Tests.cs:209]: Thanks.

**Pull request merged**